### PR TITLE
feat: support temporary table

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://docs.databend.com/dev/policies/cla/

## Summary
This PR supports temporary table in mysql session.
1. The Temporary table is only visible to the session that created it, and it will be automatically dropped when the session ends, with all data being vacuumed.
2. Within a session, you can first create a Temporary table, then create a Normal Table with the same name; or you can first create a Normal Table, then create a Temporary table with the same name. In both scenarios, the Normal Table will be hidden by the Temporary table until the Temporary table is dropped.
3. Creating a Temporary table and performing operations on it do not require any priviledges.

The syntex is:
```sql
CREATE [ OR REPLACE ] [ TRANSIENT | TEMPORARY | TEMP ] TABLE [ IF NOT EXISTS ] [ <database_name>. ]<table_name>
(
    <column_name> <data_type> [ NOT NULL | NULL ] 
                              [ { DEFAULT <expr> } ] 
                              [ AS (<expr>) STORED | VIRTUAL ]
                              [ COMMENT '<comment>' ],
    <column_name> <data_type> ...
    ...
)
```

Next step:
1. Support in http handler
2. Manually vacuum temporary table data that fails to be vacuumed automatically.
3. system.temporary_tables

## Tests

- [ ] Unit Test
- [x] Logic Test
- [ ] Benchmark Test
- [ ] No Test - _Explain why_

## Type of change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New Feature (non-breaking change which adds functionality)
- [ ] Breaking Change (fix or feature that could cause existing functionality not to work as expected)
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Performance Improvement
- [ ] Other (please describe):

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/16250)

### CODE REVIEW

File: src/query/sql/src/planner/binder/ddl/account.rs [hunk 1]
```diff
@@ -152,6 +152,15 @@ impl Binder {
                 let database_name = database_name
                     .clone()
                     .unwrap_or_else(|| self.ctx.get_current_database());
+                if self
+                    .ctx
+                    .is_temp_table(&catalog_name, &database_name, table_name)
+                {
+                    return Err(ErrorCode::StorageOther(format!(
+                        "{} is a temporary table, cannot grant privileges on it",
```
Comment (TCeason) [inline src/query/sql/src/planner/binder/ddl/account.rs:160]: {}.{}.{}, catalog, db. table_name

File: src/query/sql/src/planner/binder/ddl/account.rs [hunk 2]
```diff
@@ -197,6 +206,15 @@ impl Binder {
                 let database_name = database_name
                     .clone()
                     .unwrap_or_else(|| self.ctx.get_current_database());
+                if self
+                    .ctx
+                    .is_temp_table(&catalog_name, &database_name, table_name)
+                {
+                    return Err(ErrorCode::StorageOther(format!(
+                        "{} is a temporary table, cannot revoke privileges on it",
```
Comment (TCeason) [inline src/query/sql/src/planner/binder/ddl/account.rs:214]: {}.{}.{}, catalog, db. table_name

File: src/query/service/src/sessions/query_ctx.rs [hunk 3]
```diff
@@ -1329,8 +1330,15 @@ impl TableContext for QueryContext {
         todo!()
     }
 
-    fn is_temp_table(&self, database_name: &str, table_name: &str) -> bool {
-        todo!()
+    fn is_temp_table(&self, catalog_name: &str, database_name: &str, table_name: &str) -> bool {
+        catalog_name == CATALOG_DEFAULT
```
Comment (TCeason) [inline src/query/service/src/sessions/query_ctx.rs:1339]: Question: only catalog_default has temp table?

File: src/query/service/src/interpreters/access/privilege_access.rs [hunk 4]
```diff
@@ -937,6 +948,9 @@ impl AccessChecker for PrivilegeAccess {
 
             }
             Plan::RenameTable(plan) => {
+                if  self.ctx.is_temp_table(&plan.catalog,&plan.database, &plan.table) {
```
Comment (TCeason) [inline src/query/service/src/interpreters/access/privilege_access.rs:951]: If a user does not  have any privilege, could it drop/truncate/alter the temp table?

Maybe you could create a new user that without any privilege and then test on local with these test.
image

File: src/query/service/src/sessions/query_ctx.rs (hunk 3 above)
Comment (SkyFan2002) [reply src/query/service/src/sessions/query_ctx.rs:1339]: Yes

File: src/query/service/src/interpreters/access/privilege_access.rs (hunk 4 above)
Comment (SkyFan2002) [reply src/query/service/src/interpreters/access/privilege_access.rs:951]: When drop/truncate/alter the temp table, it should pass the table privilege check here: https://github.com/SkyFan2002/databend/blob/dc8519184446c877a15ca3ebe0cb8adb7c930a95/src/query/service/src/interpreters/access/privilege_access.rs#L296

File: src/query/sql/src/planner/binder/ddl/account.rs (hunk 1 above)
Comment (SkyFan2002) [reply src/query/sql/src/planner/binder/ddl/account.rs]: Done

File: src/query/sql/src/planner/binder/ddl/account.rs (hunk 2 above)
Comment (SkyFan2002) [reply src/query/sql/src/planner/binder/ddl/account.rs]: Done

File: src/query/service/src/interpreters/access/privilege_access.rs (hunk 4 above)
Comment (TCeason) [reply src/query/service/src/interpreters/access/privilege_access.rs:951]: > When drop/truncate/alter the temp table, it should pass the table privilege check here: https://github.com/SkyFan2002/databend/blob/dc8519184446c877a15ca3ebe0cb8adb7c930a95/src/query/service/src/interpreters/access/privilege_access.rs#L296

Ok. Now about privilege LGTM. Please add some privilege check about tmp table.

File: src/query/service/src/interpreters/interpreter_table_create.rs [hunk 5]
```diff
@@ -193,6 +194,8 @@ impl CreateTableInterpreter {
         req.table_meta.drop_on = Some(Utc::now());
         let table_meta = req.table_meta.clone();
         let reply = catalog.create_table(req.clone()).await?;
+        println!("create table as select");
+        println!("{:?}", self.ctx.session_state().temp_tbl_mgr);
         if !reply.new_table && self.plan.create_option != CreateOption::CreateOrReplace {
```
Comment (dantengsky) [inline src/query/service/src/interpreters/interpreter_table_create.rs:199]: forget to remove the `println!` or replace it with `log`?

**Comment by Xuanwo** [comment]
Hi, what's the difference between `TRANSIENT | TEMPORARY | TEMP`?

**Comment by SkyFan2002** [comment]
> Hi, what's the difference between `TRANSIENT | TEMPORARY | TEMP`?

`TEMP` is abbreviation of `TEMPORARY`, which means create a local table which lifetime and scope is limited to a single session. `TRANSIENT`  means create a table which only reserves one snapshot.

File: src/query/storages/common/table_meta/src/meta/utils.rs [hunk 6]
```diff
@@ -47,3 +55,24 @@ pub fn monotonically_increased_timestamp(
     }
     timestamp
 }
+
+pub fn parse_storage_prefix(options: &BTreeMap<String, String>, table_id: u64) -> Result<String> {
+    // if OPT_KE_STORAGE_PREFIX is specified, use it as storage prefix
+    if let Some(prefix) = options.get(OPT_KEY_STORAGE_PREFIX) {
+        return Ok(prefix.clone());
+    }
```
Comment (dantengsky) [inline src/query/storages/common/table_meta/src/meta/utils.rs:63]: Is stmt that creates temp table with external location prohibited?
, e.g. `create temp table t (..) 's3://somebuck/..`.

Just a reminder, we can address this in a dedicated PR (if necessary)

Comment (dantengsky) [reply src/query/storages/common/table_meta/src/meta/utils.rs:63]: oh... please ignore this, `OPT_KEY_STORAGE_PREFIX` is for `attache table`, nothing to do with create table with external location

**Review by TCeason — APPROVED** [review]
Privilege part is LGTM

**Comment by BohuTANG** [comment]
Conflicting files
src/query/service/src/sessions/session_ctx.rs

**Review by dantengsky — APPROVED** [review]
LGTM :+1:  supports the memory engine later? if it makes sense.

**Comment by SkyFan2002** [comment]
> LGTM 👍 supports the memory engine later? if it makes sense.

Got it.

**Comment by BohuTANG** [comment]
> cc @soyeric128 need doc

Not ready, there are still some tasks: https://github.com/databendlabs/databend/issues/16310

**Pull request merged**